test_application/src/main.rs

```diff
@@ -0,0 +1,272 @@
+                private_key_share = Some(private_key_share_local);
+            }
+            SubProtocol::StartAuxInfo => {
+                let participant: Participant<AuxInfoParticipant> =
+                    Participant::from_config(&config, sid, ());
+                // TODO: Aux info takes no arguments?!
+                let (aux_info_public_local, aux_info_private_local) =
+                    worker_handle_subprotocol(participant, &other_workers, &outgoing)?;
+                outgoing.send(MessageFromWorker::SubProtocolEnded)?;
+                aux_info_public = Some(aux_info_public_local);
+                aux_info_private = Some(aux_info_private_local);
+            }
+            SubProtocol::StartPresign => {
+                let input: PresignInput = PresignInput::new(
+                    aux_info_public.clone().unwrap(),
+                    aux_info_private.clone().unwrap(),
+                    public_key_shares.clone().unwrap(),
+                    private_key_share.clone().unwrap(),
+                )?;
+                let participant: Participant<PresignParticipant> =
+                    Participant::from_config(&config, sid, input);
+                let presign_record_local =
+                    worker_handle_subprotocol(participant, &other_workers, &outgoing)?;
+                outgoing.send(MessageFromWorker::SubProtocolEnded)?;
+                presign_record = Some(presign_record_local);
+            }
+            SubProtocol::Sign => {
+                // TODO Weird that we create a vacuous participant for signing.
+                let signature =
+                    Participant::<KeygenParticipant>::sign(presign_record.take().unwrap(), hash())?;
+                info!("Computed signature: {:?}", signature);
+                outgoing.send(MessageFromWorker::SubProtocolEnded)?;
+            }
+        }
+    }
+
+    Ok(())
+}
+
+fn hash() -> Sha256 {
+    let mut hasher = Sha256::new();
+    hasher.update(b"hello world");
+    hasher
+}
+
+#[instrument(skip_all)]
+fn worker_handle_subprotocol<P: ProtocolParticipant>(
+    mut participant: Participant<P>,
+    incoming: &Receiver<Message>,
+    outgoing: &Sender<MessageFromWorker>,
+) -> anyhow::Result<P::Output> {
+    let rng = &mut thread_rng();
+
+    debug!("Starting sub-protocol.");
+    let init_message = participant.initialize_message();
+
+    // TODO: Why does processing workers require rng?
+    let (output, messages) = participant.process_single_message(&init_message, rng)?;
+    assert!(output.is_none());
+    outgoing.send(MessageFromWorker::FinishedRound(messages))?;
+
+    for m in incoming {
+        trace!("Received message from another worker.");
+        let (output, messages) = participant.process_single_message(&m, rng)?;
+        outgoing.send(MessageFromWorker::FinishedRound(messages))?;
+
+        // TODO: Is status useful?
+
+        // Output is only Some when sub-protocol is done.
+        if let Some(output) = output {
```

Comment:
Instead of using `status` I just end up checking if I have some output to see if we finished.